Added MailAddressGenerator and tests.

cbowdon opened the pull request:
This implements the enhancement in #339.

Let me know if there's any changes you need me to make to this. I'm sorry about the newline in AutoFixtureUnitTest.csproj. It's an artifact of manually editing the file to remove some cruft Xamarin Studio added (I'm not on Win at home). I did try and remove it, but gave up after an hour or so of fiddling around (life's too short, sorry).

moodmosaic commented on Src/AutoFixture/MailAddressGenerator.cs:
```diff
@@ -0,0 +1,44 @@
+namespace Ploeh.AutoFixture
+{
+    /// <summary>
+    /// Creates new <see cref="Uri"/> instances.
```
Does this class creates new *Uri* instances?

cbowdon replied:
Ah, shameful mistake. Fixed.

moodmosaic commented on Src/AutoFixture/MailAddressGenerator.cs:
```diff
@@ -0,0 +1,44 @@
+    /// </summary>
+    public class MailAddressGenerator : ISpecimenBuilder
+    {
+        private readonly string[] fictitiousDomains = 
```
While these predefined domains correctly follow [this](https://github.com/AutoFixture/AutoFixture/issues/339#issuecomment-70984249) comment by @ploeh, wouldn't it be more flexible if we could also allow users to supply their own domains if they need to?

A similar example is the `UriGenerator` class, which [requests a `UriScheme`] (https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/UriGenerator.cs#L31) from the context.

moodmosaic commented on Src/AutoFixture/MailAddressGenerator.cs:
```diff
@@ -0,0 +1,44 @@
+        };
+
+        /// <summary>
+        /// Creates a new anonymous MailAddress.
```
Is it really an *anonymous* MailAddress?

moodmosaic commented on Src/AutoFixtureUnitTest/MailAddressGeneratorTests.cs:
```diff
@@ -0,0 +1,78 @@
+            // Fixture setup
+            var sut = new MailAddressGenerator();
+            // Exercise system
+            var dummyContext = new DelegatingSpecimenContext();
```
Is `dummyContext` part of the [Exercise SUT](http://xunitpatterns.com/exercise%20SUT.html) phase?

moodmosaic commented on Src/AutoFixtureUnitTest/MailAddressGeneratorTests.cs:
```diff
@@ -0,0 +1,78 @@
+            var request = new object();
+            var sut = new MailAddressGenerator();
+            // Exercise system
+            var dummyContext = new DelegatingSpecimenContext();
```
Is `dummyContext` part of the [Exercise SUT](http://xunitpatterns.com/exercise%20SUT.html) phase?

moodmosaic commented:
Thank you for your contribution! I just left a couple of comments, but overall (IMO) this is looking good.

It would be nice to have a [Facade Test](http://blog.ploeh.dk/2012/06/27/FacadeTest/), demoing the feature like the other tests in the [FixtureTest](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixtureUnitTest/FixtureTest.cs) class.

ecampidoglio commented on Src/AutoFixture/MailAddressGenerator.cs:
```diff
@@ -0,0 +1,44 @@
+                return new NoSpecimen(request);
+            }
+
+            var name = Guid.NewGuid().ToString();
```
If a builder needs to create a specimen it doesn't handle itself, it should usually delegate its creation to the `ISpecimenContext`. In this case, the name portion of the email address should be created by:

```csharp
var name = context.Resolve(typeof(string));
```

moodmosaic replied:
Yes, similar to [this](https://github.com/AutoFixture/AutoFixture/pull/340#discussion_r23566830) comment.

cbowdon commented:
Prompt review, thanks! I've pushed the most trivial fixes and will work on the facade test and user-supplied domains.

Regarding the UriScheme example - can you explain how it works if the user doesn't supply their own? It looks to me like the context should return null and hence the UriGenerator should return NoSpecimen.

moodmosaic commented:
The [UriGenerator](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/UriGenerator.cs) requests a [UriScheme](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/UriScheme.cs) from the context, so that users don't have to set a UriScheme, unless they want a custom one.

With the above pieces, it's possible to supply a custom URI scheme as shown below:
```c#
var fixture = new Fixture();
fixture.Inject(new UriScheme("ftp"));
var uri = fixture.Create<Uri>(); 
// Prints -> ftp://abc9f406-16f2-4e06-b6f9-0750dc115ac3/
```

This is handled by [UriSchemeGenerator](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/UriSchemeGenerator.cs) which is in turn [registered in DefaultPrimitiveBuilders iterator](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/DefaultPrimitiveBuilders.cs#L30).

—

Hypothetically, with MailAddress it should be possible to supply a custom domain name as shown below:

```c#
var fixture = new Fixture();
fixture.Inject(new DomainName("bar.baz"));
var uri = fixture.Create<MailAddress>(); 
// Prints -> abc9f406-16f2-4e06-b6f9-0750dc115ac3@bar.baz
```

Please note that it'd be wise to have @ploeh's opinion before you make this additional effort.

HTH

ploeh commented on Src/AutoFixtureUnitTest/MailAddressGeneratorTests.cs:
```diff
@@ -0,0 +1,77 @@
+            var sut = new MailAddressGenerator();
+            var dummyRequest = new object();
+            // Exercise system and verify outcome
+            Assert.DoesNotThrow(() => sut.Create(dummyRequest, null));
```
Let's say that we take `MailAddressGenerator` as is (as I'm very close to thinking that we should), but later want to enhance it so that it'll request names and domain names from the context.

If that's the case, the `context` becomes a required parameter. If we keep this test around, it might be interpreted as a breaking change. What if we didn't have this test?

ploeh commented:
This Pull Request (PR) is looking good.

Both @ecampidoglio and @moodmosaic have great suggestions, and I think that both of these suggestions should indicate the direction in which this feature should move. However, let's assume that we take this PR as it is (preferably with one very small change that I've commented elsewhere), couldn't we add both of those features later, without making them breaking changes?

cbowdon commented:
@ecampidoglio - I will flatten commits once you're happy to accept this.

moodmosaic commented:
It's possible to take this PR as it is, and add, hypothetically, DomainName and DomainNameGenerator at some later point. AFAICT, if we do that there won't be any breaking changes.

cbowdon commented:
Thanks for the feedback all. As per @ploeh's suggestion I've removed the null context test, and I believe all the earlier issues around comments are now fixed.

I have not:
- tackled the DomainName enhancement
- added a facade test to the FixtureTest class

Is this acceptable without the FixtureTest? If so I'll flatten the commits.

moodmosaic commented:
Sounds good. – With this PR, what happens if you do

```c#
var fixture = new Fixture();
var sut = fixture.Create<MailAddress>();
```
?

Wouldn't it be nice to keep track of this new behavior by adding a test in the [FixtureTest](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixtureUnitTest/FixtureTest.cs) class?

ploeh commented:
Running `build-release.sh` reports 20 build errors. Could you please address those?

ploeh commented:
If you don't want to run `build-release.sh` (which requires you to use Bash), you can also run `BuildRelease.ps1` from PowerShell; they do the same thing.

cbowdon commented:
I've pushed a couple of fixes, but please don't spend any further time on this yet. I need to clone this branch on a Windows machine and try running build-release for myself. (FWIW - trying it with xbuild on OS X resulted in a huge stream of errors unrelated to my changes.)

cbowdon commented:
I believe this is ready now. Should I update my branch though, I notice it's fallen behind master?

moodmosaic commented:
Running `build-release.sh` now completes successfully, reporting 0 warnings and 0 errors.

![image](https://cloud.githubusercontent.com/assets/287532/6112130/b81a73dc-b096-11e4-9acf-0602dde9c508.png)

ploeh commented:
Thank you for your contribution! It's now live as [AutoFixture 3.24.0](http://www.nuget.org/packages/AutoFixture/3.24.0).

Sorry it took some time before I was able to pick this up, but I've been quite snowed under these last weeks :grimacing:

cbowdon commented:
No problem, thank you all for your patience with my screw-ups!

ploeh commented:
No worries :smile: 

Are you [CBowdon](https://twitter.com/CBowdon) on Twitter? Usually, I tweet whenever there's a new minor release of AutoFixture, and drop the name of the contributor...

cbowdon commented:
Yes that's me, thanks :)

ploeh commented:
[Done](https://twitter.com/ploeh/status/568710595769462785) :smile:
